### DISCUSSION TITLE
After upgrading to axum 0.6, all my post methods are broken

### DISCUSSION THREAD

**Discussion by Moicen** [tokio-rs/axum#1755] [Q&A] [answered]
As the title side, the post methods will generate compile error like this:

```
error[E0277]: the trait bound `fn(HtyHostHeader, axum::Json<WxLogin>, axum::extract::State<Arc<DbState>>) -> impl Future<Output = axum::Json<HtyResponse<WxSession>>> {wx_login}: Handler<_, _, _>` is not satisfied
    --> htyuc/src/lib.rs:5039:44
     |
5039 |         .route("/api/v1/uc/wx_login", post(wx_login))
     |                                       ---- ^^^^^^^^ the trait `Handler<_, _, _>` is not implemented for fn item `fn(HtyHostHeader, axum::Json<WxLogin>, axum::extract::State<Arc<DbState>>) -> impl Future<Output = axum::Json<HtyResponse<WxSession>>> {wx_login}`
     |                                       |
     |                                       required by a bound introduced by this call
     |
     = help: the following other types implement trait `Handler<T, S, B>`:
               <Layered<L, H, T, S, B, B2> as Handler<T, S, B2>>
               <MethodRouter<S, B> as Handler<(), S, B>>
note: required by a bound in `post`
    --> /Users/weli/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.6.6/src/routing/method_routing.rs:407:1
     |
407  | top_level_handler_fn!(post, POST);
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `post`
     = note: this error originates in the macro `top_level_handler_fn` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Here is one of my post methods:

```rust
pub async fn wx_login(
    host: HtyHostHeader,
    Json(in_wx_login): Json<WxLogin>,
    State(db_pool): State<Arc<DbState>>,
) -> Json<HtyResponse<WxSession>> {
    debug!(":::wx_login:::");
    match raw_wx_login(&host, &in_wx_login, db_pool).await {
        Ok(ok) => wrap_json_ok_resp(ok),
        Err(e) => wrap_json_anyhow_err(e),
    }
}
```

Nothing changed in the code(just migrate to the new `State` extrator in parameter) and under `axum 0.6` it can not build.

Can anyone provide some help on this problem? Thanks!

**Comment by davidpdrsn** [comment, marked as answer]
https://docs.rs/axum/latest/axum/extract/index.html#the-order-of-extractors

**Reply by Moicen** [reply]
After changing the parameter order, and put the Json body into last position according to the doc, the problem is fixed. Thanks for help!

**Comment by bj-edward** [comment]
async fn create_todo_handler(
    claims: Claims,
    Json(todo): Json<CreateTodo>,
    Extension(store): Extension<TodoStore>,
) -> Result<StatusCode, HttpError> {
    match store.items.write() {
        Ok(mut guard) => {
            let todo = Todo {
                id: get_next_id(),
                user_id: claims.id,
                title: todo.title,
                completed: false,
            };
            guard.push(todo);
            Ok(StatusCode::CREATED)
        }
        Err(_) => Err(HttpError::Internal),
    }
}

After changing the parameter order, the problem is still there. It works well for 0.5.

**Reply by davidpdrsn** [reply]
`Json` needs to be last, like the link I posted above documents.

**Reply by bj-edward** [reply]
Thx for your quick response, I adjust the following, same result.

async fn create_todo_handler(
    claims: Claims,
    Extension(store): Extension<TodoStore>,
    Json(todo): Json<CreateTodo>,
) -> Result<StatusCode, HttpError> {
    match store.items.write() {
        Ok(mut guard) => {
            let todo = Todo {
                id: get_next_id(),
                user_id: claims.id,
                title: todo.title,
                completed: false,
            };
            guard.push(todo);
            Ok(StatusCode::CREATED)
        }
        Err(_) => Err(HttpError::Internal),
    }
}

**Reply by davidpdrsn** [reply]
Try https://docs.rs/axum/0.6.20/axum/attr.debug_handler.html
